Clarity support with pipeline for data processing

mickp opened the pull request:
Addresses issue #121. Needs testing with hardware.

carandraug commented on microscope/devices.py:
```diff
@@ -253,7 +253,10 @@ def initialize(self):
+        try:
+            self.disable()
+        except Exception as e:
+            _logger.debug("Exeption in disable() during shutdown: %s" % e)
```
We are effectively ignoring an error during disable. I guess this should at least be of warning level and not debug?

mickp replied:
We're ignoring and error during disable because there's nothing else we can do - we're shutting it down. Warning is probably the more appropriate level.

mickp replied:
Fixed - log level changed to warn in 78fe9c8b9eb8d50bbd0fd0d2060cdd63417774ed.

carandraug commented on microscope/testsuite/devices.py:
```diff
@@ -250,6 +261,7 @@ def initialize(self):
         """
         _logger.info('Initializing.')
         time.sleep(0.5)
+        self._initialized = True
```
Why not add this attribute during object construction? This would mean that one only neds to check for it's value and not have always also check if the attribute already exists.  Also, should it not be set back to `False` after shutdown?

mickp replied:
Perhaps not a good reason, but this would still work if a device fails to
set that member during __init__. In the TestCamera case, it's a simple
attribute. If we used a similar idea on real hardware, we might want some
other test (with library calls) to test that the devices is correctly
initialized, so _initialized could then be a property.

carandraug replied:
This change is only being made to `TestCamera` so there is no what other device could do.

mickp replied:
Yes, because I've introduced it here for now, but we might want to move it over to the base Device class if it's useful.

mickp commented:
No good reason, other than this would probably still work if some other
class failed to super().__init__().  But I guess it should be doing that
anyway.

On Sun, 15 Dec 2019 at 17:19, Carnë Draug <notifications@github.com> wrote:

> *@carandraug* commented on this pull request.
> ------------------------------
>
> In microscope/testsuite/devices.py
> <https://github.com/MicronOxford/microscope/pull/124#discussion_r357993000>
> :
>
> > @@ -250,6 +261,7 @@ def initialize(self):
>          """
>          _logger.info('Initializing.')
>          time.sleep(0.5)
> +        self._initialized = True
>
> Why not add this attribute during object construction? This would mean
> that one only neds to check for it's value and not have always also check
> if the attribute already exists. Also, should it not be set back to False
> after shutdown?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/MicronOxford/microscope/pull/124?email_source=notifications&email_token=ABHGTL3BGVJXUAXWUQTEFP3QYZRL7A5CNFSM4JV5D3M2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCPHCGKY#pullrequestreview-332276523>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABHGTL5G4NQBVQMCCMAQH63QYZRL7ANCNFSM4JV5D3MQ>
> .
>


-- 
________________________________
Mick Phillips
________________________________

mickp commented:
Perhaps not a good reason, but this would still work if a device fails to
set that member during __init__. In the TestCamera case, it's a simple
attribute. If we used a similar idea on real hardware, we might want some
other test (with library calls) to test that the devices is correctly
initialized, so _initialized could then be a property.

On Sun, 15 Dec 2019 at 17:19, Carnë Draug <notifications@github.com> wrote:

> *@carandraug* commented on this pull request.
> ------------------------------
>
> In microscope/testsuite/devices.py
> <https://github.com/MicronOxford/microscope/pull/124#discussion_r357993000>
> :
>
> > @@ -250,6 +261,7 @@ def initialize(self):
>          """
>          _logger.info('Initializing.')
>          time.sleep(0.5)
> +        self._initialized = True
>
> Why not add this attribute during object construction? This would mean
> that one only neds to check for it's value and not have always also check
> if the attribute already exists. Also, should it not be set back to False
> after shutdown?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/MicronOxford/microscope/pull/124?email_source=notifications&email_token=ABHGTL3BGVJXUAXWUQTEFP3QYZRL7A5CNFSM4JV5D3M2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCPHCGKY#pullrequestreview-332276523>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABHGTL5G4NQBVQMCCMAQH63QYZRL7ANCNFSM4JV5D3MQ>
> .
>


-- 
________________________________
Mick Phillips
________________________________

carandraug commented on microscope/filterwheels/aurox.py:
```diff
@@ -157,11 +166,14 @@ def _c_process_data(self, data):
-        return self._devices
+    def devices(self) -> Mapping[str, microscope.devices.Device]:
+        """Devices property, required by ControllerDevice interface."""
+        return {'camera': self._cam}
```
If there is no `CameraDevice` this should return an empty dict, and not a camera with a `None` value. Also, the possible devices of a controller should be documented.

mickp replied:
Returning ```{}``` indicates "I don't have any devices", but ```{'camera': None}``` indicates I should have a camera but don't". Is this more useful, or maybe a different use case? I guess this applies more to the Clarity because it's not much use without a camera, whereas other controllers may aggregate a bunch of optional devices (e.g. a stage controller that may have a variable number of axes attached).

mickp replied:
Where should the documentation go?

carandraug replied:
We are specifically trying to avoid having `None`s as meaning that it could have that device but does not. See issue #111 but basically it's tricky to derive what the full list of possible devices is. Examples: having a device connected may prevent others from being connected (the filterwheel 3 on prior proscan prevents having a light source; the ZaberDaisyChain controller can have up to 255 devices connected.

On the documentation about what controlled devices may exist, it should go on the docstring of the class that implements the Controller interface.

mickp replied:
c5e5270: Clarity now returns {} if no camera instead of {'camera': None}

mickp replied:
Docstrings added in 0c98ba9.

carandraug commented on microscope/devices.py:
```diff
@@ -667,18 +673,19 @@ def __init__(self, **kwargs):
+               (0, 1): numpy.flipud(numpy.rot90(data, rot)),
+               (1, 0): numpy.fliplr(numpy.rot90(data, rot)),
+               (1, 1): numpy.fliplr(numpy.flipud(numpy.rot90(data, rot)))
+               }[flips]
```
Not introduced in this change but isn't this indexing of the dict always perfoming all flips and rotations and then we are discarding them all except one? If so, shouldn't we ensure that we only do the transformation that is actually required?

mickp replied:
Good point. I assumed lazy evaluation, but maybe that's not the case. Perhaps lambdas would be more appropriate.

mickp replied:
It's not evaluated lazily.  I'll wrap it with lambda and push a commit to the master branch.

mickp replied:
The fix is in this branch, because it'll introduce a merge conflict here if I push it to master.

mickp replied:
Fix included in this PR: cfff4a590a5b94b5aa5e680f05ebb6cf989d1bf0.

carandraug commented on microscope/devices.py:
```diff
@@ -481,8 +483,12 @@ def _fetch_data(self):
+        Subclasses should call super()._process_data(data) after doing their
+        own processing."""
+        import functools
+        return functools.reduce(lambda x, f: f(x), self.pipeline, data)
```
If we have an attribute which is a list of processing steps, why should subclasses do their own processing first instead of just appending to `self.pipeline`?

mickp replied:
Yes - I think we should do one or the other. This requires some thought, though.

Originally ```_process_data``` was a method on ```DataDevice```s to implement the processing step, and I think that was meant to be pure virtual.

Now we have ```Devices``` that are not ```DataDevices``` that need to add processing steps, and I addressed that with the pipeline, but that needs ```super()._process_data()``` to ensure that the pipeline is executed after the ```DataDevice``` does anything it needs to do (e.g. running it through a LUT, type conversions, reshaping arrays). 

The ```DataDevice``` could just add it's processing to the start of the pipeline, instead, and ```_process_data``` would be implemented on ```DataDevice``` and just process the pipeline.

mickp replied:
We talked about this at length. 
* Most classes derived from ```DataDevice``` should probably add to the pipeline rather than override ```_process_data```.
* Devices that override ```_process_data``` must call ```super()._process_data()```.
* The docstring on ```DataDevice._process_data``` should indicate this clearly.
* We could make ```pipeline``` private, but that would require methods to add/remove steps from it.

mickp replied:
See comments in review immediately below this one.

carandraug commented on microscope/devices.py:
```diff
@@ -409,6 +409,8 @@ def __init__(self, buffer_length=0, **kwargs):
         # A condition to signal arrival of a new data and unblock grab_next_data
         self._new_data_condition = threading.Condition()
+        # A data processing pipeline: a list of f(data) -> data.
+        self.pipeline = []
```
Should we not make this a private attribute? Also, would be nice to also have type annotations for it.

mickp replied:
Maybe it should be private, but then we'd need methods to allow other devices to add their processing steps.

mickp replied:
We talked about this at length.
* Most (all?) classes derived from DataDevice should probably add to the pipeline rather than override _process_data.
* Devices that override _process_data must call super()._process_data().
* The docstring on DataDevice._process_data should indicate this clearly.
* We could make pipeline private, but that would require methods to add/remove steps from it.

mickp replied:
```_process_data``` docstring made more explicit in 1c16359.
```CameraDevice``` made to use ```pipeline``` instead of override in 0361a02.
```_process_data``` is not only defined in one place, on the ```DataDevice``` class.

carandraug commented on microscope/filterwheels/aurox.py:
```diff
@@ -86,16 +97,44 @@ class Clarity(microscope.devices.FilterWheelBase):
+        cam_kwargs = {}
+        for key in cam_kw_keys:
+            cam_kwargs[key.replace("camera.", "")] = kwargs[key]
+            del kwargs[key]
```
Would a `camera_kwargs` argument with the arguments to construct the camera make this cleaner?

    def __init__(self, camera=None, camera_kwargs={}, **kwargs):
        super().__init__(**kwargs)
        if camera is not None:
            self._camera = camera(**camera_kwargs)
            ...

mickp replied:
It might make the code a bit cleaner, but then you'd have nested ```{ {} }``` in the config, and maybe that's not cleaner / more error-prone.

carandraug commented:
I have spoken with Mick earilier today in person about this PR and the
following is just a summary of that.

This Clarity mixing a `ControllerDevice` with a `FilterWheel` should
really be some sort of composite device.  The problem it tries to
solve by being a controller is the same that we have on AO-tools and
with live SIM reconstruction (at least in the manner that Marcel
envisioned it).

Not mentioned anywhere yet on this PR is that the reasoning behind
this design is that we want to avoid passing data around between
processes for processing: camera gets image, sends it to the Clarity
for processing, Clarity then sends it back to the camera, which
finally sends it to the client.  Using a controller makes that pretty
easy since everything ends up in the same process.

However, we are just delaying the problem and we will eventually need
to come up with some interface to composite devices and possibly how
to share data between processes (the multiprocessing module has
classes for that but there's a whole bunch of issues that we may have
to think about, specially since we mainly want to shared numpy arrays
and not python arrays).

For now, we need this Aurox interface and it does do what we need it
to do.  It enables us to delay coming up with a general solution to
composite devices.

mickp commented:
I've addressed some of the review items, and rebased this onto master.
I think the only outstanding issues are:
 * whether or not to make ```pipeline``` private
 * whether to use ```camera.some_parameter=x``` or ```camera_kwargs = {'some_parameter': x}```

carandraug commented:
> whether to use camera.some_parameter=x or camera_kwargs = {'some_parameter': x}

It just occurred to me that we can't use a dot on parameter name, the following will fail because dot can't be used in a keyword:

    Clarity(camera.some_parameter=x)

even though this works:

    Clarity(**{'camera.some_parameter':x})

mickp commented:
So it currently only works because the ```DeviceServer``` uses the second
method to call the constructor?

On Mon, 6 Jan 2020 at 11:25, Carnë Draug <notifications@github.com> wrote:

> whether to use camera.some_parameter=x or camera_kwargs =
> {'some_parameter': x}
>
> It just occurred to me that we can't use a dot on parameter name, the
> following will fail because dot can't be used in a keyword:
>
> Clarity(camera.some_parameter=x)
>
> even though this works:
>
> Clarity(**{'camera.some_parameter':x})
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/MicronOxford/microscope/pull/124?email_source=notifications&email_token=ABHGTL37QU5AIMBZNDHPU3DQ4MIMBA5CNFSM4JV5D3M2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEIFF4LI#issuecomment-571104813>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABHGTL5YB2FE7NBTBNPD3F3Q4MIMBANCNFSM4JV5D3MQ>
> .
>


-- 
________________________________
Mick Phillips
________________________________

carandraug commented:
> So it currently only works because the ```DeviceServer``` uses the second
> method to call the constructor?

Yes.

This point is just about the use of a period on an argument name, it's not about whether passing a dict of arguments to the camera vs passing multiple args marked for the camera.

mickp commented:
I think we need another approach for this.

Data processing can modify the shape of returned arrays, yet when we query this, the request goes to the underlying data source which returns the original shape. With the pipeline approach on one method, it's not easy to address this. A better approach might be to take the underlying camera class, dynamically create an augmented class that modifies both the data collection and data shape methods, and serve an instance of that augmented class.

carandraug commented:
We have #133 which is a different implementation also from Mick, that addresses some of the issues from this solution. We are not merging this. Closing.
